components/youtube/sources/new-videos-by-username/new-videos-by-username.js

```diff
@@ -1,84 +1,57 @@
-const youtube = require("../../youtube.app.js");
+const common = require("../common.js");
 
 module.exports = {
+  ...common,
   key: "youtube-new-videos-by-username",
   name: "New Videos by Username",
   description: "Emits an event for each new Youtube video tied to a username.",
-  version: "0.0.1",
+  version: "0.0.2",
   dedupe: "unique",
   props: {
-    youtube,
+    ...common.props,
     username: {
       type: "string",
       label: "Username",
       description: "Search for new videos uploaded by the YouTube Username.",
     },
-    db: "$.service.db",
-    timer: {
-      type: "$.interface.timer",
-      default: {
-        intervalSeconds: 60 * 15,
-      },
+  },
+  methods: {
+    ...common.methods,
+    getParams(channelId) {
+      return {
+        part: "snippet",
+        type: "video",
+        order: "date",
+        channelId,
+      };
     },
   },
-
   async run(event) {
-    let videos = [];
-    let channelIds = [];
-    let totalResults;
-    let nextPageToken;
-    let count;
-    let results;
-
-    const now = new Date();
-    const monthAgo = now;
-    monthAgo.setMonth(monthAgo.getMonth() - 1);
-    const publishedAfter = this.db.get("publishedAfter") || monthAgo;
+    let publishedAfter = this._getPublishedAfter();
+    let lastPublished;
 
-    let params = {
+    const channelParams = {
       part: "id",
       forUsername: this.username,
     };
 
-    channels = await this.youtube.getChannels(params);
-    channels.data.items.forEach(function (channel) {
-      channelIds.push(channel.id);
+    const channels = (await this.youtube.getChannels(channelParams)).data;
+    const channelIds = channels.items.map((channel) => {
+      return channel.id;
     });
 
     for (const channelId of channelIds) {
-      count = 0;
-      totalResults = 1;
-      nextPageToken = null;
-      params = {
-        part: "snippet",
-        type: "video",
-        channelId: this.channelId,
-        pageToken: null,
-        publishedAfter,
-      };
-
-      while (count < totalResults) {
-        params.pageToken = nextPageToken;
-        results = await this.youtube.getVideos(params);
-        totalResults = results.data.pageInfo.totalResults;
-        nextPageToken = results.data.nextPageToken;
-        results.data.items.forEach(function (video) {
-          videos.push(video);
-          count++;
-        });
-        if (!nextPageToken) break;
-        if (!results.data.items || results.data.items.length < 1) break;
-      }
+      const params = this.getParams(channelId);
+      if (publishedAfter) params.publishedAfter = publishedAfter;
+      else params.maxResults = 10;
```

Comment:
Is this logic in place so that we send the initial events?
Because since we're overriding the `run()` method here, we kind of know that `publishedAfter` will be set, unless it's the very first execution.

I think it might be cleaner to have this logic in the hooks of the `common` module perhaps?
```js
// sources/common.js

module.exports = {
  hooks: {
    async deploy() {
      const params = {
        ...this._getBaseParams(),
        maxResults: 10,
        ...this.getParams(),
      };
      // Paginate through videos...
    },
    async activate() {
      this._setPublishedAfter(Date.now());
    },
  },
  methods: {
    _getBaseParams() {
      return {
        part: "snippet",
        type: "video",
        order: "date",
      };
    },
  },
  async run() {
    const publishedAfter = this._getPublishedAfter();
    const params = {
      ...this._getBaseParams(),
      publishedAfter,
      ...this.getParams(),
    };
    // Do the rest...
  },
};
```

Reply:
The purpose of this PR is to limit the videos pulled on the first run.

> On the initial run of the trigger, we attempt to fetch a number of sample videos for the specified user, and the user you specified had so many videos that we hit the YouTube quota just from trying to list those sample videos.
> 
> We're optimizing how we fetch these initial videos and should have an approach that keeps you under limits. I'll let you know when we have a new version of the trigger for you to test.

The idea here is to add in maxResults on the first run in order to limit the videos pulled and try to avoid going over Youtube's quota limit. I moved that logic to a deploy hook in common.js and overwrote it for components that don't need it or need different logic in the deploy.